packages/eslint-plugin/src/rules/no-invalid-void-type.ts

```diff
@@ -0,0 +1,119 @@
+import {
+  AST_NODE_TYPES,
+  TSESTree,
+} from '@typescript-eslint/experimental-utils';
+import * as util from '../util';
+
+const getFullyQualifiedName = (node: TSESTree.EntityName): string => {
+  if (node.type === AST_NODE_TYPES.Identifier) {
+    return node.name;
+  }
+
+  return `${getFullyQualifiedName(node.left)}.${node.right.name}`;
+};
+
+interface Options {
+  allowGenerics: boolean | string[];
+}
+
+type MessageIds =
+  | 'invalidVoidForGeneric'
+  | 'invalidVoidNotReturnOrGeneric'
+  | 'invalidVoidNotReturn';
+
+export default util.createRule<[Options], MessageIds>({
+  name: 'no-invalid-void-type',
+  meta: {
+    type: 'problem',
+    docs: {
+      description:
+        'Disallows usage of `void` type outside of generic or return types',
+      category: 'Best Practices',
+      recommended: false,
+    },
+    messages: {
+      invalidVoidForGeneric:
+        '{{ generic }} may not have void as a type variable',
+      invalidVoidNotReturnOrGeneric:
+        'void is only valid as a return type or generic type variable',
+      invalidVoidNotReturn: 'void is only valid as a return type',
+    },
+    schema: [
+      {
+        type: 'object',
+        properties: {
+          allowGenerics: {
+            oneOf: [
+              { type: 'boolean' },
+              {
+                type: 'array',
+                items: { type: 'string' },
+                minLength: 1,
+              },
+            ],
+          },
+        },
+        additionalProperties: false,
+      },
+    ],
+  },
+  defaultOptions: [{ allowGenerics: true }],
+  create(context, [{ allowGenerics }]) {
+    const validParents: AST_NODE_TYPES[] = [
+      AST_NODE_TYPES.TSTypeAnnotation, //
+    ];
+    const invalidGrandParents: AST_NODE_TYPES[] = [
+      AST_NODE_TYPES.TSPropertySignature,
+      AST_NODE_TYPES.CallExpression,
+      AST_NODE_TYPES.ClassProperty,
+      AST_NODE_TYPES.Identifier,
+    ];
+
+    if (allowGenerics === true) {
+      validParents.push(AST_NODE_TYPES.TSTypeParameterInstantiation);
+    }
+
+    return {
+      TSVoidKeyword(node: TSESTree.TSVoidKeyword): void {
+        /* istanbul ignore next */
+        if (!node.parent?.parent) {
+          return;
+        }
+
+        if (
+          validParents.includes(node.parent.type) &&
+          !invalidGrandParents.includes(node.parent.parent.type)
+        ) {
+          return;
+        }
+
+        if (
+          node.parent.type === AST_NODE_TYPES.TSTypeParameterInstantiation &&
+          node.parent.parent.type === AST_NODE_TYPES.TSTypeReference &&
+          Array.isArray(allowGenerics)
+        ) {
+          const fullyQualifiedName = getFullyQualifiedName(
+            node.parent.parent.typeName,
+          );
```

Comment:
> Also important to note - you should remove spaces from the input type, or else there might be weird happenings like:

You mean like doing `.replace(/ /gu, '')`?

Happy to do that, but that example case feels to me like the weird thing thats happening: if `allowGenerics` takes exact matches, who are helping by supporting "almost exact matches except for whitespaces".

Again, don't mind add support, but just wondering if I'm missing a use case 😅

Reply:
> You mean like doing .replace(/ /gu, '')?

Yup! See code from `ban-types`:
https://github.com/typescript-eslint/typescript-eslint/blob/b01f5e778ac28e0797a3734fc58d025bb224f418/packages/eslint-plugin/src/rules/ban-types.ts#L134-L136

It's good just to make sure there's no edge cases we miss due to spacing. In case someone accidentally adds a space to the end of a string, or does some formatting in the string like something weird `TypeA[ 'property' ]`

Reply:
ahhh right you mean the type - I thought you meant the options property :joy:

Reply:
sorry - I did mean the options property as well.
removing spaces from both means there can't ever be any mismatch.